### ISSUE TITLE
Analyzer crashes with 'System.ArgumentOutOfRangeException'

### ISSUE THREAD

**Issue by genlu** [dotnet/roslyn#20330]
Happens in Project System repo, after switching to latest analyzers package, the exception is triggered by project `Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests`.

```
Detail Description
Analyzer driver threw the following exception:'System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.Parameter name: specialType   
at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSpecialType(SpecialType specialType)   
at Microsoft.CodeAnalysis.CSharp.LocalRewriter.GetDefaultParameterValue(SyntaxNode syntax, ParameterSymbol parameter, ThreeState enableCallerInfo, LocalRewriter localRewriter, Binder binder, DiagnosticBag diagnostics)   
at Microsoft.CodeAnalysis.CSharp.LocalRewriter.AppendMissingOptionalArguments(SyntaxNode syntax, Symbol methodOrIndexer, MethodSymbol optionalParametersMethod, Boolean expanded, Binder binder, ArrayBuilder`1 missingParameters, ArrayBuilder`1 argumentsBuilder)   
at Microsoft.CodeAnalysis.CSharp.LocalRewriter.BuildArgumentsInEvaluationOrder(SyntaxNode syntax, Symbol methodOrIndexer, MethodSymbol optionalParametersMethod, Boolean expanded, ImmutableArray`1 argsToParamsOpt, ImmutableArray`1 arguments, Binder binder)   
at Microsoft.CodeAnalysis.CSharp.LocalRewriter.MakeArgumentsInEvaluationOrder(Binder binder, SyntaxNode syntax, ImmutableArray`1 arguments, Symbol methodOrIndexer, MethodSymbol optionalParametersMethod, Boolean expanded, ImmutableArray`1 argsToParamsOpt, Boolean invokedAsExtensionMethod)  
at Microsoft.CodeAnalysis.CSharp.BoundCall.<>c__DisplayClass26_0.<DeriveArguments>b__0(BoundExpression n)   
at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValue(TKey key, CreateValueCallback createValueCallback)   
at Microsoft.CodeAnalysis.CSharp.BoundCall.DeriveArguments(BoundExpression boundNode, Binder binder, Symbol methodOrIndexer, MethodSymbol optionalParametersMethod, ImmutableArray`1 boundArguments, ImmutableArray`1 argumentNamesOpt, ImmutableArray`1 argumentsToParametersOpt, ImmutableArray`1 argumentRefKindsOpt, ImmutableArray`1 parameters, Boolean expanded, SyntaxNode invocationSyntax, Boolean invokedAsExtensionMethod)   
at Microsoft.CodeAnalysis.CSharp.BoundCall.Microsoft.CodeAnalysis.Semantics.IHasArgumentsExpression.get_ArgumentsInEvaluationOrder()   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.VisitInvocationExpression(IInvocationExpression operation)   
at Microsoft.CodeAnalysis.CSharp.BoundCall.Accept(OperationVisitor visitor)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.Visit(IOperation operation)   
at Microsoft.CodeAnalysis.Semantics.OperationExtensions.OperationCollector.Visit(IOperation operation)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.VisitVariableDeclaration(IVariableDeclaration operation)   
at Microsoft.CodeAnalysis.Semantics.VariableDeclaration.Accept(OperationVisitor visitor)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.Visit(IOperation operation)   
at Microsoft.CodeAnalysis.Semantics.OperationExtensions.OperationCollector.Visit(IOperation operation)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.VisitArray[T](ImmutableArray`1 list)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.VisitVariableDeclarationStatement(IVariableDeclarationStatement operation)   
at Microsoft.CodeAnalysis.CSharp.BoundLocalDeclaration.Accept(OperationVisitor visitor)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.Visit(IOperation operation)   
at Microsoft.CodeAnalysis.Semantics.OperationExtensions.OperationCollector.Visit(IOperation operation)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.VisitArray[T](ImmutableArray`1 list)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.VisitBlockStatement(IBlockStatement operation)  
at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(OperationVisitor visitor)   
at Microsoft.CodeAnalysis.Semantics.OperationWalker.Visit(IOperation operation)  
at Microsoft.CodeAnalysis.Semantics.OperationExtensions.OperationCollector.Visit(IOperation operation)  
at Microsoft.CodeAnalysis.Semantics.OperationExtensions.DescendantsAndSelf(IOperation operation)  
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.GetOperationsToAnalyze(ImmutableArray`1 operationBlocks)  
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.TryExecuteDeclaringReferenceActions(SyntaxReference decl, Int32 declarationIndex, SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisStateOpt, Boolean shouldExecuteSyntaxNodeActions, Boolean shouldExecuteOperationActions, Boolean shouldExecuteCodeBlockActions, Boolean shouldExecuteOperationBlockActions, Boolean isInGeneratedCode, CancellationToken cancellationToken)-----'.
```

**Comment by genlu** [comment]
@dotnet/analyzer-ioperation

**Comment by genlu** [comment]
Turns out the cause of this is the compilation we get to run analyzers on has 0 references, probably still being constructed by underlying project system. As a result, for something like below, the type of the parameter x would be `ErrroType`, then we ends up getting type symbol for `SpecialType.None`.

```cs
class P
{
    static void M1()
    {
        M2();
    }

    static void M2(bool? x = null)
    {
    }
}
```

The fix I'm proposing is simple, instead of checking `BoundNode.HasErrors` before deriving arguments (which I thought would be sufficient to filter out all error cases), we check `BoundNode.HasAnyErrors` (so we bail if there's any error on types/syntax as well, later this will become the `InvalidInvocationOperation` when fixing #18722).
